src/foam/u2/FormattedTextField.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.u2',
+  name: 'FormattedTextField',
+  extends: 'foam.u2.View',
+
+  implements: [
+    'foam.mlang.Expressions'
+  ],
+
+  imports: [
+  ],
+
+  css: `
+    ^ {
+      display: flex;
+    }
+
+    ^container-selection {
+      display: flex;
+      justify-content: center;
+      align-items: center;
+      text-align: center;
+
+      box-sizing: border-box;
+      width: 64px;
+      height: 30px;
+
+      border: 1px solid /*%GREY3%*/ #cbcfd4;
+      border-right: none;
+      border-radius: 3px 0 0 3px;
+    }
+
+    ^container-selection p {
+      margin: 0;
+    }
+
+    ^container-input {
+      box-sizing: border-box;
+      flex: 1;
+      height: 30px;
+
+      font-size: 14px;
+
+      border: 1px solid /*%GREY3%*/ #cbcfd4;
+      border-left: none;
+      border-radius: 0 3px 3px 0;
+    }
+  `,
+
+  properties: [
+    {
+      name: 'prop'
+    },
+    {
+      class: 'String',
+      name: 'valueString',
+      documentation: 'This is the front facing formatted value',
+      factory: function() {
+        return this.data || '';
+      },
+      preSet: function(_, v) {
+        var sanitized = this.sanitizeString(v);
+        this.data = sanitized;
+        var formatted = this.prop.tableCellFormatter.f(
+          sanitized, this.prop);
+        return formatted;
+      },
+      view: {
+        class: 'foam.u2.tag.Input',
+        onKey: true
+      }
+    }
+  ],
+
+  methods: [
+    function initE() {
+      this.SUPER();
+      var self = this;
+      this.addClass(this.myClass())
+        .add(this.slot(function(mode) {
+          if ( mode === foam.u2.DisplayMode.RW ) {
+            return this.E().style({ 'display': 'flex' })
+            .startContext({ data: self })
+              .start(self.VALUE_STRING).addClass(self.myClass('container-input'))
+              .end()
+            .endContext();
+          }
+        }));
+    },
+    function fromProperty(prop) {
+      this.SUPER(prop);
+
+      this.prop = prop;
+    },
+    // TODO: move this to property of foam.core.String
+    function sanitizeString(s) {
+      return s.replace(/[\D\s\._\-\/]+/g, "");
```

Comment:
This would remove anything that isn't a number. Is this the behaviour you want for a generalized FormattedTextView?

Reply:
tbh I assumed `\D` was for unicode dashes but I just looked it up and it's non-digits. I'll remove that token - thanks for the catch!

Reply:
A property property called `sanitize` would be great for this because then you can override the answer to "what's a formatting character?" on any property, but I'll wait for @kgrgreer 's input on this

Reply:
I like the idea of being able to provide a regex for it. It definitely makes it more dynamic.

However, I am wondering about the side effects in regards to UX. The CurrencyInputView has a side effect of moving the cursor back to the end if you inserted a value midway and it reformats. It's not a major thing for number input since I intended it to behave more like an atm or calculator, but I wonder about other things. Just food for thought.

Reply:
I think I know an elegant way to fix all of that - adding `[DNM]` tag, I'll commit a fix soon

Reply:
Got it working pretty well now

Reply:
So playing around with it some more, it's probably good enough but has some undesirable quirks like deleting forwards (like PC del key or fn+delete on osx) while at a syntax character moves the cursor in the wrong spot. Also the code I wrote to make it work is a little complicated which is kind of a double-edged sword - like I feel like there's gotta be a much simpler way to do this but at the same time now we have a neat string diff utility if I factor it out.